[tools/depends] bump cmake 3.20.3 + xcode new build system support

fuzzard opened the pull request:
## Description
Bump cmake to 3.20.3
Fix support for Xcode "new" build system for the osx target, but also explicitly set xcode project to continue to use the "legacy" system for all apple targets

## Motivation and context
Cmake 3.19+ in combination with xcode 12+ by default will use Xcode's "new" build system. There are supposed benefits, however in my testing the speed-ups suggested arent really apparent for our codebase. Apple have indicated they will remove the "legacy" build system at some point in the future.

The "new" xcode build system was introduced in Xcode 10, and cmake now believe that Xcode 12 now works to a sufficient level with a cmake generated project and the "new" system.

For our purposes i think we should continue to use the legacy build system for now (CI uses xcode 11 where cmake will default to the legacy system anyway), as this allows consistency for users who build themselves and need to compare to the official releases. 

The final commit actually makes changes to support a "new" build system project. Without this, the new system fails (change in env vars available and weird cmake quoting issues). In local testing the "legacy" system continued to work with these changes.

IOS new build system
The ios/tvos targets with the new build system are still not fully functional. Cmake appears to have some issues generating a project that uses add_custom_command(OUTPUT) as an input to a target. For some reason the scripts created for the new build system arent being built in order (parallelisation) as a dependency, and are therefore sometimes not being run prior to the target, and then build failures occur. executing the target a second time always succeeds. There are 2 targets this happens with (python_bindings, compileinfo). 
I was dealing with this issue with cmake 3.18 + new build system for osx, but bumping to 3.20 "solved" the issue. May be able to wait on cmake to fix it upstream, or the better solution may be to alter our cmake to specifically imply the dependency. A problem for another time though. 

## How has this been tested?
Locally using Xcode 12.4 on Catalina 10.15.7 against OSX target. ios legacy target still works, but the new build system ios target is not fully functional as yet.

## What is the effect on users?
Should be no effect for users, as we are choosing to actively use the same build system prior to this PR

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[tools/depends][native] Update cmake to 3.20.3

Updates cmake to 3.20.3
Set buildsystem for xcode generated projects to use the legacy build system to stay inline
with existing behaviour. Cmake 3.19+ when used with xcode 12+ will try to default to the
"new" xcode build system

Commit:
[xcode] fixes to allow xcode 10+ "new" build system to work

kambala-decapitator commented on cmake/scripts/darwin_embedded/Macros.cmake:
```diff
@@ -2,7 +2,9 @@ function(core_link_library lib wraplib)
-    set(wrapper_obj cores/dll-loader/exports/kodi.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/wrapper.build/Objects-$(CURRENT_VARIANT)/$(CURRENT_ARCH)/wrapper.o)
+    # CURRENT_VARIANT is an Xcode env var
+    # CPU is a project cmake var
+    set(wrapper_obj cores/dll-loader/exports/kodi.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/wrapper.build/Objects-$CURRENT_VARIANT/${CPU}/wrapper.o)
```
why removing parentheses in `CURRENT_VARIANT`? (also in osx)

fuzzard replied:
yeah, so thats the quoting issue i allude to. for an xcode buildphase script, the xcode env vars are accessible via $VARIABLE_NAME. however, configuration and effective_platform_name are actually unique in that cmake explicitly check for $(CONFIGURATION) and strip and pass it as $CONFIGURATION into the generated script. No other env var is handled this way for whatever reason. Cmake does some weird escaping at times (cmake issue for reference/explanation https://gitlab.kitware.com/cmake/cmake/-/issues/18062)

For some reason the new xcode build system fails (where the old one doesnt care). 

Also the change to CPU is because CURRENT_ARCH is no longer available. ARCHS can be a string of multiple (eg x86_64:arm64), and the wrap_libdvdnav target doesnt have an arch so we just have to rely on our cmake var for that.

kambala-decapitator replied:
that's quite some black magic involved... Thank you for the detailed explanation!

kambala-decapitator reviewed (approved):
looks good, just one unclear point

Commit:
[tools/depends][target] mariadb fix cmake 3.20 build error

Build failure in mariadb-connector-c

 checking how to run the C preprocessor... CMake Error at cmake/ConnectorName.cmake:30 (ENDMACRO):
   Flow control statements are not properly nested.
 Call Stack (most recent call first):
   CMakeLists.txt:434 (INCLUDE)

This fix is already upstream in mariadb-connector-c 3.1.13. Patch can be dropped when we
bump to a newer version

wsnipex commented on tools/depends/target/mariadb/Makefile:
```diff
@@ -23,6 +23,7 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
 	cd $(PLATFORM); patch -p1 -i ../01-android.patch
 	cd $(PLATFORM); patch -p1 -i ../02-symbol_rename-gnutls_clash.patch
+	cd $(PLATFORM); patch -p1 -i ../03-cmake3.20_syntax_fix.patch
```
did you forget to add that patch?

fuzzard replied:
Yeah, just woke up and out all morning, I'll get it fixed up this afternoon

fuzzard commented:
Ive updated the PR description. Currently we have issues with ios/tvos targets with the new build system. Before this i had already decided for reasons to stick with the old one for now anyway, so not something i have to fix right now.

@wsnipex can you chuck the cmake 3.20.3 tarball up on mirrors when you get time.

wsnipex commented on tools/depends/native/cmake/Makefile:
```diff
@@ -3,10 +3,12 @@ PLATFORM=$(NATIVEPLATFORM)
+SOURCE=$(APPNAME)-v$(VERSION)
 ARCHIVE=$(SOURCE).tar.gz
 
+BASE_URL=https://gitlab.kitware.com/cmake/cmake/-/archive/v$(VERSION)
```
uploaded cmake-3.20.3.tar.gz
